### PR TITLE
Allow third party methods to appear in local pickup area

### LINKED ISSUE #8012: Support third party shipping rates in local pickup
As it stands right now, only core local pickup (id `pickup_location`) is using the pickup flow, which is fine and all but it's kind of limiting in a sense that third party shipping providers can only use half the shipping flow (regular ones) but not the local pickup flow.

We should provide some sort of filter that contains an array of shipping method ids, all of those methods will be shown alongside local pickup.

Right now, everything is hardcoded to `pickup_location`.

### PR DESCRIPTION
**Pull request by opr**
This PR will allow third party shipping methods that support 'collectible' to be displayed under local pickup.

To achieve this, I created a method in `ShippingController.php` to get all registered methods that declare `collectible` support, then add their IDs to the `collectibleMethodIds` key in the asset data registry.

This can be retrieved on the client by `getSetting< string[] >( 'collectibleMethodIds', [] )`.

When deciding which methods to show in the `checkout-pickup-options` block it now checks to see if the method name is in this `collectibleMethodIds` array **or** if the method id is `pickup_location`.

It also removes methods, following the same rules as above, from being displayed in the "regular" shipping options area.

I also modified `assets/js/base/context/hooks/shipping/types.ts` to set `packageId` as optional, this type definition did not correctly reflect the implementation.

This PR also removes hardcoded references to `pickup_location` when it is used to determine whether a local pickup method is in use, and instead uses the `collectibleMethodIds` site setting.

Fixes #8012

#### Automated Tests
* [x] Changes in this PR are covered by Automated Tests.
  * [x] Unit tests
  * [ ] E2E tests

#### Internal developer testing

1. Add the following code somewhere on your site, it will register a test shipping method called "Amazon locker".

<details>
<summary>Expand for example shipping method</summary>

```php
function amazon_locker_shipping_init() {
	class Amazon_Locker_Shipping_Method extends WC_Shipping_Method {

		/**
		 * Min amount to be valid.
		 *
		 * @var integer
		 */
		public $min_amount = 0;

		/**
		 * Requires option.
		 *
		 * @var string
		 */
		public $requires = '';

		/**
		 * Constructor.
		 *
		 * @param int $instance_id Shipping method instance.
		 */
		public function __construct( $instance_id = 0 ) {
			$this->id                 = 'amazon_locker_shipping';
			$this->instance_id        = absint( $instance_id );
                        $this->title              = 'Amazon locker';
			$this->method_title       = __( 'Amazon locker', 'woocommerce' );
			$this->method_description = __( 'Get your order shipped to an amazon locker.', 'woocommerce' );
			$this->supports           = array(
				'instance-settings',
				'instance-settings-modal',
				'local-pickup',
			);

			$this->init();
		}

		/**
		 * Initialize Amazon Locker shipping.
		 */
		public function init() {
		}

		/**
		 * See if Amazon locker shipping is available based on the package and cart.
		 *
		 * @param array $package Shipping package.
		 * @return bool
		 */
		public function is_available( $package ) {
			return true;
		}

		/**
		 * Called to calculate shipping rates for this method. Rates can be added using the add_rate() method.
		 *
		 * @param array $package Shipping package.
		 * @uses WC_Shipping_Method::add_rate()
		 */
		public function calculate_shipping( $package = array() ) {
			$this->add_rate(
				array(
					'label'   => $this->title,
					'cost'    => 0,
					'taxes'   => false,
					'package' => $package,
				)
			);
		}
	}
}

add_action( 'woocommerce_shipping_init', 'amazon_locker_shipping_init' ); // use this hook to initialize your new custom method

function add_amazon_locker_shipping( $methods ) {
	$methods['amazon_locker_shipping'] = 'Amazon_Locker_Shipping_Method';

	return $methods;
}
add_filter( 'woocommerce_shipping_methods', 'add_amazon_locker_shipping' );
```
</details>

2. Enable this method for one of your shipping zones. Note, the settings UI doesn't work properly for this method but that is OK.
3. Enable the "normal" local pickup too, add a location.
4. Add some items to your cart, go to the Cart page, enter an address using the shipping calculator. Ensure you see the local pickup option and the amazon locker option.
5. Go to the Checkout and, without selecting local pickup, scroll down to the shipping methods section, you should not see amazon locker there.
6. Choose local pickup, and ensure the amazon locker option appears.

#### User Facing Testing

1. Enable local pickup on your site.
2. Add items to your cart and go to the Cart block.
3. Add an address using the Shipping Calculator. Ensure you see the local pickup options you added.
4. Go to the Checkout page and, **without** selecting local pickup, scroll down to the shipping options area. Ensure you do not see local pickup there.
8. Select local pickup and ensure the pickup location is shown.
9. Check out successfully.

* [ ] Do not include in the Testing Notes 

### WooCommerce Visibility

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Changelog

> Allow third party shipping methods to declare compatibility with WC Blocks local pickup.

### CODE REVIEW

File: assets/js/base/context/hooks/shipping/types.ts [hunk 1]
```diff
@@ -12,7 +12,7 @@ export interface ShippingData {
 	// Returns a function that accepts a shipping rate ID and a package ID.
 	selectShippingRate: (
 		newShippingRateId: string,
-		packageId: string | number
+		packageId?: string | number | undefined
```
Comment (mikejolley) [inline assets/js/base/context/hooks/shipping/types.ts:15]: 👍🏻 noticed this too

File: src/Shipping/ShippingController.php [hunk 2]
```diff
@@ -60,6 +61,25 @@ function() {
...
+	 *
+	 * @return string[] List of payment method ids that support the 'collectible' feature.
+	 */
+	public function get_collectible_method_ids() {
+		return array_keys(
+			wp_list_pluck(
+				array_filter(
+					WC()->shipping()->get_shipping_methods(),
+					function( $method ) {
+						return in_array( 'collectible', $method->supports, true );
```
Comment (mikejolley) [inline src/Shipping/ShippingController.php]: Did we avoid [this method](https://github.com/woocommerce/woocommerce/blob/3611d4643791bad87a0d3e6e73e031bb80447417/plugins/woocommerce/includes/abstracts/abstract-wc-shipping-method.php#L154-L156) for a reason?

File: src/Shipping/ShippingController.php [hunk 3]
```diff
@@ -60,6 +61,25 @@ function() {
 		add_filter( 'pre_update_option_pickup_location_pickup_locations', array( $this, 'flush_cache' ) );
 	}
 
+	/**
+	 * Gets a list of payment method ids that support the 'collectible' feature.
+	 *
+	 * @return string[] List of payment method ids that support the 'collectible' feature.
+	 */
+	public function get_collectible_method_ids() {
+		return array_keys(
```
Comment (mikejolley) [inline src/Shipping/ShippingController.php]: Is there a logic error here? 

`wp_list_pluck` would give you just the IDs from the filtered methods, but the indexes would be numeric. `array_keys` would just take those indexes right?

We could use `array_values`, or alternatively simplify this to use `array_map`—make the callback return either the method ID if its `collectible`, or return `false`. Then you can just run `array_filter` over the resulting mapped array.

File: src/Shipping/ShippingController.php (hunk 2 above)
Comment (opr) [reply src/Shipping/ShippingController.php]: I'm using it now, thanks for showing me it.

File: src/Shipping/ShippingController.php (hunk 3 above)
Comment (opr) [reply src/Shipping/ShippingController.php]: Updated now, no longer using pluck, and also inserting `pickup_location` here so we don't have to in the client.

File: src/Shipping/ShippingController.php [hunk 4]
```diff
@@ -60,6 +61,27 @@ function() {
...
+	 * Gets a list of payment method ids that support the 'collectible' feature.
+	 *
+	 * @return string[] List of payment method ids that support the 'collectible' feature.
+	 */
+	public function get_collectible_method_ids() {
+		$collectible_method_ids = array_keys(
+			array_filter(
+				WC()->shipping()->get_shipping_methods(),
+				function( $method ) {
+					return $method->supports( 'collectible' );
```
Comment (mikejolley) [inline src/Shipping/ShippingController.php]: I'd like to suggest we make this wording/variable slightly clearer by making it:

`return $method->supports( 'collection' );`

or

`return $method->supports( 'local_pickup' );`

What do you think?

File: assets/js/blocks/checkout/inner-blocks/checkout-shipping-method-block/shared/helpers.ts [hunk 5]
```diff
@@ -2,12 +2,18 @@
...
  */
 import type { CartShippingPackageShippingRate } from '@woocommerce/type-defs/cart';
+import { getSetting } from '@woocommerce/settings';
 
 export interface minMaxPrices {
 	min: CartShippingPackageShippingRate | undefined;
 	max: CartShippingPackageShippingRate | undefined;
 }
 
+const collectibleMethodIds = getSetting< string[] >(
```
Comment (mikejolley) [inline assets/js/blocks/checkout/inner-blocks/checkout-shipping-method-block/shared/helpers.ts]: Nothing wrong here, however, we could probably move this settings call to a new shared util which returns a function called `isRateCollectable` or similar. This can be consumed here and in assets/js/blocks/checkout/inner-blocks/checkout-pickup-options-block/block.tsx

Comment (mikejolley) [reply assets/js/blocks/checkout/inner-blocks/checkout-shipping-method-block/shared/helpers.ts]: If this method accepted either a string or array of strings, you could also replace `areRatesCollectible` with this helper. The code that splits `rate.split( ':' )[ 0 ]` method ids would remain in place. Thoughts?

File: src/Shipping/ShippingController.php [hunk 6]
```diff
@@ -60,6 +61,27 @@ function() {
...
+				},
+				true
+			)
+		);
+		// `pickup_location` will always be collectible. Adding it here means we can avoid hard coding it elsewhere on
+		// the client.
+		$collectible_method_ids[] = 'pickup_location';
```
Comment (mikejolley) [inline src/Shipping/ShippingController.php]: Why would the code above not return pickup_location?

**Review by mikejolley — APPROVED** [review]
Works great, the test extension worked as intended. I've made a few suggestions for shared utils/naming inline. They are not blocking but I think the `supports` variable could use something else. I'll leave it with you.

File: src/Shipping/ShippingController.php (hunk 4 above)
Comment (opr) [reply src/Shipping/ShippingController.php]: Sounds good, I will update it.

File: src/Shipping/ShippingController.php [hunk 7]
```diff
@@ -60,6 +61,30 @@ function() {
 		add_filter( 'pre_update_option_pickup_location_pickup_locations', array( $this, 'flush_cache' ) );
 	}
 
+	/**
+	 * Gets a list of payment method ids that support the 'local_pickup' feature.
+	 *
+	 * @return string[] List of payment method ids that support the 'local_pickup' feature.
+	 */
+	public function get_local_pickup_method_ids() {
+		$methods_supporting_local_pickup = array_unique(
```
Comment (opr) [inline src/Shipping/ShippingController.php]: I had to go back to doing it this way as `WC()->shipping()->get_shipping_methods()` returns like this:

image

For some reason it can include the same method multiple times, so stripping out duplicates and only outputting the method ID seems most sensible.

Is there a cleaner way of doing this?

File: src/Shipping/ShippingController.php (hunk 6 above)
Comment (opr) [reply src/Shipping/ShippingController.php]: It does now, because of this change: https://github.com/woocommerce/woocommerce-blocks/pull/8256/files#diff-b9026989674ea1e4ec8f06e05634d871425015639a2177cd79499cb8ed564d1fR29

File: assets/js/blocks/checkout/inner-blocks/checkout-shipping-method-block/shared/helpers.ts (hunk 5 above)
Comment (opr) [reply assets/js/blocks/checkout/inner-blocks/checkout-shipping-method-block/shared/helpers.ts]: See changes in `assets/js/base/utils/shipping-rates.ts`, this just checks a rate, or rates, and doesn't take care of splitting

File: src/Shipping/ShippingController.php (hunk 7 above)
Comment (mikejolley) [reply src/Shipping/ShippingController.php]: How about this:

```php
		$methods_supporting_local_pickup = array_reduce(
			WC()->shipping()->get_shipping_methods(),
			function( $methods, $method ) {
				if ( $method->supports( 'local_pickup' ) ) {
					$methods[] = $method->id;
				}
				return $methods;
			},
			array()
		);
```

Comment (opr) [reply src/Shipping/ShippingController.php]: Nice, that looks better, but I still need to `array_unique` it, given `WC()->shipping()->get_shipping_methods()` can return the same one multiple times. I'll implement.

**Review by mikejolley — APPROVED** [review]
Still testing well 👍🏻 Let's ship it.

File: assets/js/base/context/hooks/shipping/types.ts [hunk 8]
```diff
@@ -12,10 +12,12 @@ export interface ShippingData {
...
 	) => void;
 	// Only true when ALL packages support local pickup. If true, we can show the collection/delivery toggle
 	isCollectable: boolean;
 	// True when a rate is currently being selected and persisted to the server.
 	isSelectingRate: boolean;
+
+	hasSelectedLocalPickup: boolean;
```
Comment (senadir) [inline assets/js/base/context/hooks/shipping/types.ts:22]: Why is this needed?

File: assets/js/blocks/checkout/inner-blocks/checkout-pickup-options-block/block.tsx [hunk 9]
```diff
@@ -118,7 +119,7 @@ const Block = (): JSX.Element | null => {
 
 	// Get pickup locations from the first shipping package.
 	const pickupLocations = ( shippingRates[ 0 ]?.shipping_rates || [] ).filter(
-		( { method_id: methodId } ) => methodId === 'pickup_location'
+		isPackageRateCollectable
```
Comment (senadir) [inline assets/js/blocks/checkout/inner-blocks/checkout-pickup-options-block/block.tsx:122]: `isPackageRateCollectable` and `hasCollectableRate` seem to handle similar logic, why can't we use just one?

File: src/Shipping/PickupLocation.php [hunk 10]
```diff
@@ -26,6 +26,7 @@ public function init() {
 		$this->title            = $this->get_option( 'title' );
 		$this->tax_status       = $this->get_option( 'tax_status' );
 		$this->cost             = $this->get_option( 'cost' );
+		$this->supports         = [ 'local-pickup' ];
```
Comment (senadir) [inline src/Shipping/PickupLocation.php:29]: This should be documented, we currently don't have a dev documentation page for Local Pickup, but once we do, we should include this. Also this PR needs a dev note for this one.

**Review by senadir — COMMENTED** [review]
Overall this is a great PR, I left some small notes that we might address later.
